### PR TITLE
osu-lazer{,-bin}: 2024.906.2 -> 2024.1009.1, refactor

### PR DESCRIPTION
**Pull request by gepbird**
Release: https://github.com/ppy/osu/releases/tag/2024.1009.1

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by gepbird** [comment]
Update was pulled due to crashes.

Also it was marked as a pre-release, and the in-game update checker said an older version is the latest:
image

If this is intentional, we should adjust the update script to exclude pre-releases.

**Comment by gepbird** [comment]
The issue was fixed and a new version was released.

File: pkgs/games/osu-lazer/bin.nix [hunk 1]
```diff
@@ -1,31 +1,34 @@
...
+        url = "https://github.com/ppy/osu/releases/download/${version}/osu.app.Intel.zip";
+        hash = "sha256-ToxDZHL59YPmybvB9tsiOnFEd+FJJE4mNMfaK6btYKo=";
+        stripRoot = false;
+      };
+      x86_64-linux = fetchurl {
+        url = "https://github.com/ppy/osu/releases/download/${version}/osu.AppImage";
+        hash = "sha256-zQnR3KwlE1gTWH8f+GDRBsc7Whfn9XpT1D/NLg5TtrU=";
+      };
+    }
+    .${stdenv.system} or (throw "${pname}-${version}: ${stdenv.system} is unsupported.");
```
Comment (SuperSandro2000) [inline pkgs/games/osu-lazer/bin.nix]:
```suggestion
    .${stdenv.system} or (throw "osu-lazer-bin: ${stdenv.system} is unsupported.");
```
pname is not necessarily the attr which is important here.

File: pkgs/by-name/os/osu-lazer-bin/update.sh [hunk 2]
```diff
@@ -3,16 +3,16 @@
...
 
 new_version="$(curl -s "https://api.github.com/repos/ppy/osu/releases?per_page=1" | jq -r '.[0].name')"
-old_version="$(sed -nE 's/\s*version = "(.*)".*/\1/p' ./bin.nix)"
+old_version="$(sed -nE 's/\s*version = "(.*)".*/\1/p' ./package.nix)"
 if [[ "$new_version" == "$old_version" ]]; then
     echo "Already up to date."
     exit 0
 fi
 
-cd ../../..
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply pkgs/by-name/os/osu-lazer-bin/update.sh:15]: in the end, it doesn't really matter

File: pkgs/by-name/os/osu-lazer-bin/package.nix [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      ;
+
+    extraPkgs = pkgs: with pkgs; [ icu ];
+
+    extraInstallCommands =
+      let
+        contents = appimageTools.extract { inherit pname version src; };
+      in
+      ''
+        . ${makeWrapper}/nix-support/setup-hook
```
Comment (SuperSandro2000) [inline pkgs/by-name/os/osu-lazer-bin/package.nix:91]: We can't use nativeBuildInputs here?

File: pkgs/by-name/os/osu-lazer-bin/update.sh
Comment (thiagokokada) [inline pkgs/by-name/os/osu-lazer-bin/update.sh:1]: Can you test this script with:

```
nix-shell maintainers/scripts/update.nix --argstr package osu-lazer-bin
```

And make sure it works?

File: pkgs/by-name/os/osu-lazer/update.sh
Comment (thiagokokada) [inline pkgs/by-name/os/osu-lazer/update.sh:1]: Can you test this script with:

```
nix-shell maintainers/scripts/update.nix --argstr package osu-lazer
```

And make sure it works?

File: pkgs/by-name/os/osu-lazer-bin/package.nix (hunk 3 above)
Comment (gepbird) [reply pkgs/by-name/os/osu-lazer-bin/package.nix:91]: No, see https://github.com/NixOS/nixpkgs/pull/340125#discussion_r1748054515

**Review by stepbrobd — APPROVED** [review]

**Review by thiagokokada — APPROVED** [review]

**Comment by ilsubyeega** [comment]
did `https://github.com/ppy/osu/releases/download/2024.1009.1/osu.AppImage` get changed? i'm getting hash mismatch, which is `sha256-PAbLw/IVJHe/y7YSyrTK9wJVcN3VIZYhrgDoHzYtjS8=` instead of `sha256-G7+Mr0Zd/p8XUVU5aG7KywvLD2hz2N37kEbm5rlghEw=`

**Comment by gepbird** [comment]
> did `https://github.com/ppy/osu/releases/download/2024.1009.1/osu.AppImage` get changed? i'm getting hash mismatch, which is `sha256-PAbLw/IVJHe/y7YSyrTK9wJVcN3VIZYhrgDoHzYtjS8=` instead of `sha256-G7+Mr0Zd/p8XUVU5aG7KywvLD2hz2N37kEbm5rlghEw=`
> 
> edit: asked on discord, looks like republish

@ilsubyeega interesting, yesterday they made a 2024.1009.0 which had critical bugs, so they pulled it and made a new release, 2024.1009.1. I'm not aware of pulling this one too and re-releasing it with the same version, but if that's the case I'll update it in a few hours (if it's not fixed already).

**Pull request merged**